File: python/binarySearchTree/avlTree.py
from node import Node

class AvlTree:
    
    def getRoot(self, node):
        if node.parent is None:
            return node
        return self.getRoot(node.parent)

    def _find(self, key, node):
        if key == node.key:
            return node

        if key > node.key:
            if node.right is None:
                return node
            return self._find(key, node.right)
        else:
            if node.left is None:
                return node
            return self._find(key, node.left)

    def _insert(self, key, root):
        if root is None:
            return Node(key, None)

        node = self._find(key, root)
        newNode = Node(key, node)
        
        if key > node.key:
            node.right = newNode
        else:
            node.left = newNode
        
        return root

    def avlInsert(self, key, root):
        #print('AVL Insert: ', key)
        root = self._insert(key, root)
        node = self._find(key, root)
        self._rebalance(node)

        return root

    def _rebalance(self, node):
        #print('Rebalance: ', node.key, ' | ', self._getHeight(node.left) , '/', self._getHeight(node.right))
        if self._getHeight(node.left) > self._getHeight(node.right) + 1:
            self._rebalanceRight(node)

        if self._getHeight(node.right) > self._getHeight(node.left) + 1:
            self._rebalanceLeft(node)

        self._adjustHeight(node)

        p = node.parent
        if p is not None:
            self._rebalance(p)

    def _getHeight(self, node):
        return 0 if node is None else node.height

    def _rebalanceRight(self, node):
        #print('Rebalance Right: ', node.key)
        pivot = node.left
        if self._getHeight(pivot.right) > self._getHeight(pivot.left):
            self._rotateLeft(pivot)
        self._rotateRight(node)

    def _rebalanceLeft(self, node):
        #print('Rebalance Left: ', node.key)
        pivot = node.right
        if self._getHeight(pivot.left) > self._getHeight(pivot.right):
            self._rotateRight(pivot)
        self._rotateLeft(node)

    def _rotateRight(self, node):
        #print('Rotate Right: ', node.key)
        p = node.parent
        b = node.left
        e = b.right

        if p is not None:
            if p.left is not None and p.left.key == node.key:
                p.left = b
            if p.right is not None and p.right.key == node.key:
                p.right = b
        b.parent = p

        b.right = node
        node.parent = b

        node.left = e
        if e is not None:
            e.parent = node
    
    def _rotateLeft(self, node):
        #print('Rotate Left: ', node.key)
        p = node.parent
        c = node.right
        d = c.left

        if p is not None:
            if p.left is not None and p.left.key == node.key:
                p.left = c
            if p.right is not None and p.right.key == node.key:
                p.right = c
        c.parent = p

        c.left = node
        node.parent = c

        node.right = d
        if d is not None:
            d.parent = node

    def _adjustHeight(self, node):
        if node is None:
            return 0
        
        node.height = max(self._adjustHeight(node.left), self._adjustHeight(node.right)) + 1
        #print('Adjust Height: ', node.key, ' / ', node.height)
        return node.height

    def bfsPrint(self, root):
        if root is None:
            return

        result = []
        result.append(root)

        while (len(result) > 0):
            node = result.pop(0)
            print(node.key, '(', node.height, ')', end=' ')
            
            if node.left is not None:
                result.append(node.left)
            if node.right is not None:
                result.append(node.right)

    def avlRemove(self, node):
        tmp = self._remove(node)
        
        if tmp is None:
            return None

        self._rebalance(tmp)
        
        return self.getRoot(tmp)

    def _remove(self, node):
        if node.right is None:
            p = node.parent
            l = node.left

            if l is not None:
                l.parent = p
            
            if p is not None:
                if p.left is not None and p.left.key == node.key:
                    p.left = l
                if p.right is not None and p.right.key == node.key:
                    p.right = l
                return p
            else:
                return l
            
        else:
            nextNode = self._next(node)
            node.key = nextNode.key

            p = nextNode.parent
            r = nextNode.right

            if p is not None:
                if p.left is not None and p.left.key == nextNode.key:
                    p.left = r
                if p.right is not None and p.right.key == nextNode.key:
                    p.right = r

            if r is not None:
                r.parent = p

            return p

    def _next(self, node):
        if node.right is not None:
            return self._leftDecendant(node.right)
        else:
            return self._rightAncestor(node.key, node.parent)

    def _leftDecendant(self, node):
        if node.left is None:
            return node
        return self._leftDecendant(node.left)

    def _rightAncestor(self, key, node):
        if node is None:
            return None

        if key < node.key:
            return node
        else:
            return self._rightAncestor(key, node.parent)

File: java/src/binarySearchTree/AVLTree.java
package binarySearchTree;

import java.util.LinkedList;
import java.util.Queue;

public class AVLTree {
    public void rotateRight(Node node) {
        Node p = node.parent;
        Node l = node.left;
        Node r = l.right;

        if (p != null) {
            if (p.left != null && p.left.key == node.key)
                p.left = l;
            else if (p.right != null && p.right.key == node.key)
                p.right = l;
        }
        l.parent = p;

        l.right = node;
        node.parent = l;

        node.left = r;
        if (r != null)
            r.parent = node;
    }

    public void rotateLeft(Node node) {
        Node p = node.parent;
        Node r = node.right;
        Node l = r.left;

        if (p != null) {
            if (p.left != null && p.left.key == node.key)
                p.left = r;
            else if (p.right != null && p.right.key == node.key)
                p.right = r;
        }
        r.parent = p;

        r.left = node;
        node.parent = r;

        node.right = l;
        if (l != null)
            l.parent = node;
    }

    public Node find(int key, Node node) {
        if (key == node.key)
            return node;

        if (key > node.key) {
            if (node.right == null)
                return node;
            return this.find(key, node.right);
        } else {
            if (node.left == null)
                return node;
            return this.find(key, node.left);
        }
    }

    public Node insert(int key, Node root) {
        if (root == null)
            return new Node(key, null);

        Node node = this.find(key, root);
        Node newNode = new Node(key, node);

        if (key > node.key)
            node.right = newNode;
        else
            node.left = newNode;

        return root;
    }

    public Node avlInsert(int key, Node root) {
        root = this.insert(key, root);
        Node node = this.find(key, root);
        this.rebalance(node);

        return root;
    }

    public void rebalance(Node node) {
        if (this.getHeight(node.left) > this.getHeight(node.right) + 1)
            this.rebalanceRight(node);
        if (this.getHeight(node.right) > this.getHeight(node.left) + 1)
            this.rebalanceLeft(node);

        this.adjustHeight(node);

        Node p = node.parent;
        if (p != null)
            this.rebalance(p);
    }

    public void rebalanceRight(Node node) {
        Node l = node.left;
        if (this.getHeight(l.right) > this.getHeight(l.left))
            this.rotateLeft(l);
        this.rotateRight(node);
    }

    public void rebalanceLeft(Node node) {
        Node r = node.right;
        if (this.getHeight(r.left) > this.getHeight(r.right))
            this.rotateRight(r);
        this.rotateLeft(node);
    }

    public int getHeight(Node node) {
        return node == null? 0 : node.height;
    }

    public int adjustHeight(Node node) {
        if (node == null)
            return 0;
        return node.height = Math.max(this.adjustHeight(node.left), this.adjustHeight(node.right)) + 1;
    }

    public Node getRoot(Node node) {
        if (node.parent != null)
            return this.getRoot(node.parent);
        return node;
    }

    public void bfsPrint(Node node) {
        Queue<Node> q = new LinkedList();
        q.add(node);

        while(true) {
            Node tmp = q.poll();

            if(tmp == null)
                break;

            System.out.printf("%d(%d) ", tmp.key, tmp.height);

            if (tmp.left != null)
                q.add(tmp.left);
            if (tmp.right != null)
                q.add(tmp.right);
        }
        System.out.println();
    }

    public Node avlRemove(Node node) {
        Node p = this.remove(node);

        if (p == null)
            return null;

        this.rebalance(p);

        return this.getRoot(p);
    }

    public Node remove(Node node) {
        if (node.right == null) {
            Node p = node.parent;
            Node l = node.left;

            if (l != null)
                l.parent = p;
            else
                return null;

            if (p != null) {
                if (p.left != null && p.left.key == node.key)
                    p.left = l;
                if (p.right != null && p.right.key == node.key)
                    p.right = l;

                return p;
            }
            return l;
        } else {
            Node next = this.next(node);
            Node p = next.parent;
            Node r = next.right;

            node.key = next.key;

            if (r != null)
                r.parent = p;

            if (p.left != null && p.left.key == next.key)
                p.left = r;
            if (p.right != null && p.right.key == next.key)
                p.right = r;

            return p;
        }
    }

    public Node next(Node node) {
        if (node.right != null)
            return this.leftDescendant(node.right);
        else
            return this.rightAncestor(node.key, node.parent);
    }

    public Node leftDescendant(Node node) {
        if (node.left != null)
            return this.leftDescendant(node.left);
        return node;
    }

    public Node rightAncestor(int key, Node node) {
        if (node.key > key) {
            return node;
        } else {
            if (node.parent == null)
                return null;

            return this.rightAncestor(key, node.parent);
        }
    }
}
